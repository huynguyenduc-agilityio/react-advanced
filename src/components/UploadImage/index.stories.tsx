import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { StoryObj, Meta } from '@storybook/react';
import { Flex } from '@chakra-ui/react';

// Components
import UploadImage from '.';

const queryClient = new QueryClient();

const meta = {
  title: 'Components/UploadImage',
  component: UploadImage,
  parameters: {
    layout: 'centered',
  },
  render: (arg) => (
    <QueryClientProvider client={queryClient}>
      <Flex justifyContent="center" w="1000px" p={9} bgColor="spaceBlue">
        <UploadImage {...arg} />
      </Flex>
    </QueryClientProvider>
  ),
  tags: ['autodocs'],
} satisfies Meta<typeof UploadImage>;

export default meta;

type Story = StoryObj<typeof UploadImage>;

export const Default: Story = {
  args: {
    onFileChange: () => {},
  },
};

export const HasImage: Story = {
  args: {
    imageUrl: '/path-to-avatar.jpg',
  },
};
