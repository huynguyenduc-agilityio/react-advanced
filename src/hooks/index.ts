export * from './usePagination';
export * from './useUpload';
export * from './useUser';
export * from './useCustomToast';
