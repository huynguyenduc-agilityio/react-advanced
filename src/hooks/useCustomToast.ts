import { useToast } from '@chakra-ui/react';

// Types
import { ToastStatus } from '@/types';

interface ICustomToastProps {
  status: ToastStatus;
  message: string;
}

export const useCustomToast = () => {
  const toast = useToast();

  return ({ status, message }: ICustomToastProps) => {
    toast({
      title: message,
      status: status,
      position: 'top-right',
      isClosable: true,
      variant: 'left-accent',
      duration: 3000,
      containerStyle: {
        fontSize: 'sm',
      },
    });
  };
};
