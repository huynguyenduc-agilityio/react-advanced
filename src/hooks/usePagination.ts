import { useCallback } from 'react';

export const usePagination = (
  pageCount: number,
  currentPage: number,
  setCurrentPage: React.Dispatch<React.SetStateAction<number>>,
  setPageSize: React.Dispatch<React.SetStateAction<number>>,
) => {
  const isDisablePrev = currentPage <= 1;

  const isDisableNext = currentPage === pageCount;

  const handleChangePage = useCallback(
    (direction: string) => {
      setCurrentPage(direction === 'prev' ? currentPage - 1 : currentPage + 1);
    },
    [currentPage, setCurrentPage],
  );

  const handleChangeLimit = useCallback(
    (limit: number) => {
      setPageSize(limit);
      setCurrentPage(1);
    },
    [setCurrentPage, setPageSize],
  );

  return {
    isDisableNext,
    isDisablePrev,
    handleChangeLimit,
    handleChangePage,
  };
};
