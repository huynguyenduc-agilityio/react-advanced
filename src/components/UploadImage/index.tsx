import {
  memo,
  useCallback,
  useMemo,
  useRef,
  useState,
  useTransition,
} from 'react';
import { Controller, useForm } from 'react-hook-form';
import {
  Avatar,
  Flex,
  FormControl,
  FormErrorMessage,
  IconButton,
  Input,
  Text,
} from '@chakra-ui/react';

// Constants
import {
  ERROR_MESSAGES,
  SUCCESS_MESSAGES,
  VALIDATION_RULES,
} from '@/constants';

// Types
import { ToastStatus } from '@/types';

// Hooks
import { useCustomToast, useUploadImage } from '@/hooks';

// Components
import { ImageIcon } from '@/components';
import Fallback from '../Fallback';

export interface TUploadImageProps {
  imageUrl?: string;
  onFileChange: (file: string) => void;
}

const UploadImage = ({ imageUrl = '', onFileChange }: TUploadImageProps) => {
  const { handleUploadImage } = useUploadImage();
  const showToast = useCustomToast();
  const [selectedImageUrl, setSelectedImageUrl] = useState<string | null>(
    imageUrl,
  );
  const [isPending, startTransition] = useTransition();
  const fileInputRef = useRef<HTMLInputElement>(null);
  const {
    control,
    setValue,
    trigger,
    clearErrors,
    formState: { errors },
  } = useForm<{ image: File }>({
    mode: 'onBlur',
    reValidateMode: 'onChange',
  });

  const handleOpenFile = () => {
    fileInputRef.current?.click();
  };

  const handleUploadFile = useCallback(
    async (file: File) => {
      try {
        const isValid = await trigger('image');

        if (isValid) {
          const imageUrl = await handleUploadImage(file);

          startTransition(() => {
            setSelectedImageUrl(imageUrl);
            onFileChange(imageUrl);
            clearErrors('image');

            showToast({
              status: ToastStatus.Success,
              message: SUCCESS_MESSAGES.IMAGE,
            });
          });
        }
      } catch (error) {
        showToast({
          status: ToastStatus.Error,
          message: ERROR_MESSAGES.UPLOAD_IMAGE_FAILED,
        });
      }
    },
    [clearErrors, handleUploadImage, onFileChange, showToast, trigger],
  );

  const handleFileChange = useCallback(
    async (e: React.ChangeEvent<HTMLInputElement>) => {
      const file = e.target.files?.[0];

      if (file) {
        setValue('image', file, { shouldValidate: true });

        handleUploadFile(file);
      }
    },
    [handleUploadFile, setValue],
  );

  const handleRemoveImage = useCallback(() => {
    if (selectedImageUrl) {
      URL.revokeObjectURL(selectedImageUrl);
    }
    setSelectedImageUrl(null);
  }, [selectedImageUrl]);

  const renderImage = useMemo(() => {
    if (isPending) {
      return <Fallback />;
    }
    if (selectedImageUrl) {
      return (
        <Flex flexDirection="column" alignItems="center" width="48px">
          <Avatar
            data-testid="uploaded-image"
            src={selectedImageUrl}
            boxSize="48px"
            borderRadius="full"
          />
          <Text
            fontSize="xs"
            cursor="pointer"
            color="pastelBlue"
            _hover={{ color: 'white' }}
            onClick={handleRemoveImage}
          >
            Delete
          </Text>
        </Flex>
      );
    }

    return null;
  }, [handleRemoveImage, isPending, selectedImageUrl]);

  return (
    <Flex direction="column" align="start" gap={4}>
      {renderImage}
      {!selectedImageUrl && (
        <Controller
          control={control}
          rules={VALIDATION_RULES.IMAGE}
          name="image"
          render={() => (
            <FormControl isInvalid={!!errors?.image}>
              <Flex
                p={6}
                textAlign="center"
                flexDirection="column"
                alignItems="center"
                justifyContent="center"
              >
                <IconButton
                  aria-label="Upload image"
                  variant="icon"
                  icon={<ImageIcon />}
                  cursor="pointer"
                  onClick={handleOpenFile}
                />
                <Text fontSize="xs" mt={2} color="pastelBlue">
                  <Text
                    as="span"
                    color="primary"
                    cursor="pointer"
                    onClick={handleOpenFile}
                  >
                    Click to upload
                  </Text>{' '}
                  or drag and drop
                </Text>
                <Text fontSize="xs" color="pastelBlue">
                  SVG, PNG, JPG or GIF (max. 800 x 400px)
                </Text>
                <Input
                  type="file"
                  ref={fileInputRef}
                  accept="image/*"
                  onChange={handleFileChange}
                  hidden
                  id="imageInput"
                  data-testid="file-input"
                />
              </Flex>

              {errors?.image && (
                <FormErrorMessage>
                  <Flex direction="column" gap={4} color="red">
                    <Text>{errors.image.message}</Text>
                    <Text>Please select again!!!</Text>
                  </Flex>
                </FormErrorMessage>
              )}
            </FormControl>
          )}
        />
      )}
    </Flex>
  );
};

export default memo(UploadImage);
