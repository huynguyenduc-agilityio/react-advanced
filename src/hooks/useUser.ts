import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';

// Services
import {
  addUser,
  deleteMultipleUsers,
  deleteUser,
  editUser,
  getUser,
  getUsers,
} from '@/services';

// Constants
import { ERROR_MESSAGES, SUCCESS_MESSAGES, usersQueryKeys } from '@/constants';

// Enums
import { ToastStatus } from '@/enums';

// Components
import { useCustomToast } from './useCustomToast';

interface IUseUser {
  page: number;
  pageSize: number;
  filter?: Record<string, string>;
}

export const useGetUsers = ({ page, pageSize, filter }: IUseUser) => {
  const { data: users, isFetching: isUsersLoading } = useQuery({
    queryKey: [...usersQueryKeys.list({ page, pageSize, filter })] as const,
    queryFn: getUsers,
  });

  return { users, isUsersLoading };
};

export const useGetUser = (id: string) => {
  const { data: user, isFetching: isUserLoading } = useQuery({
    queryKey: [...usersQueryKeys.detail(id)] as const,
    queryFn: getUser,
  });

  return { user, isUserLoading };
};

export const useCreateUser = () => {
  const queryClient = useQueryClient();
  const showToast = useCustomToast();

  const { mutateAsync: handleCreateUser, isLoading: isCreateLoading } =
    useMutation({
      mutationFn: addUser,
      onSuccess: (response) => {
        queryClient.invalidateQueries({ queryKey: usersQueryKeys.lists() });

        if (response.id) {
          queryClient.setQueryData(
            usersQueryKeys.detail(response.id),
            response,
          );
        }

        showToast({
          status: ToastStatus.Success,
          message: SUCCESS_MESSAGES.ADD_USER,
        });
      },
      onError: () => {
        showToast({
          status: ToastStatus.Error,
          message: ERROR_MESSAGES.ADD_USER,
        });
      },
    });

  return { handleCreateUser, isCreateLoading };
};

export const useUpdateUser = () => {
  const queryClient = useQueryClient();
  const showToast = useCustomToast();

  const { mutateAsync: handleUpdateUser, isLoading: isUpdateLoading } =
    useMutation({
      mutationFn: editUser,
      onSuccess: (response) => {
        queryClient.invalidateQueries({ queryKey: usersQueryKeys.lists() });

        if (response.id) {
          queryClient.setQueryData(
            usersQueryKeys.detail(response.id),
            response,
          );
        }

        showToast({
          status: ToastStatus.Success,
          message: SUCCESS_MESSAGES.UPDATE_USER,
        });
      },
      onError: () => {
        showToast({
          status: ToastStatus.Error,
          message: ERROR_MESSAGES.UPDATE_USER,
        });
      },
    });

  return { handleUpdateUser, isUpdateLoading };
};

export const useDeleteUser = () => {
  const queryClient = useQueryClient();

  const { mutateAsync: handleDeleteUser, isLoading: isDeleteLoading } =
    useMutation({
      mutationFn: deleteUser,
      onSuccess: () =>
        queryClient.invalidateQueries({
          queryKey: usersQueryKeys.lists(),
        }),
    });

  return { handleDeleteUser, isDeleteLoading };
};

export const useDeleteMultipleUsers = () => {
  const queryClient = useQueryClient();

  const {
    mutateAsync: handleDeleteMultipleUsers,
    isLoading: isDeleteMultipleLoading,
  } = useMutation({
    mutationFn: deleteMultipleUsers,
    onSuccess: () =>
      queryClient.invalidateQueries({
        queryKey: usersQueryKeys.lists(),
      }),
  });

  return { handleDeleteMultipleUsers, isDeleteMultipleLoading };
};
