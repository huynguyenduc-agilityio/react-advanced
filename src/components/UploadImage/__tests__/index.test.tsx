import { fireEvent, render, waitFor } from '@testing-library/react';

import { useUploadImage } from '@/hooks';

import userEvent from '@testing-library/user-event';
import UploadImage from '..';

jest.mock('@/hooks', () => ({
  ...jest.requireActual('@/hooks'),
  useUploadImage: jest.fn(),
}));

describe('UploadImage Component', () => {
  const mockHandleUploadImage = jest.fn();
  const mockOnFileChange = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    (useUploadImage as jest.Mock).mockReturnValue({
      handleUploadImage: mockHandleUploadImage,
    });
  });

  it('renders the component with initial props', () => {
    const { getByText } = render(
      <UploadImage onFileChange={mockOnFileChange} />,
    );

    expect(getByText(/Click to upload/i)).toBeInTheDocument();
    expect(getByText(/or drag and drop/i)).toBeInTheDocument();
  });

  it('triggers file input when clicking on upload area', () => {
    const { getByText, getByTestId } = render(
      <UploadImage onFileChange={mockOnFileChange} />,
    );

    const fileInput = getByTestId('file-input');
    jest.spyOn(fileInput, 'click');
    fireEvent.click(getByText(/Click to upload/i));
    expect(fileInput.click).toHaveBeenCalled();
  });

  it('opens file input when clicking on the upload area', async () => {
    const { getByTestId } = render(
      <UploadImage onFileChange={mockOnFileChange} />,
    );

    const fileInput = getByTestId('file-input');
    jest.spyOn(fileInput, 'click');

    await userEvent.click(getByTestId('file-input'));
    expect(fileInput.click).toHaveBeenCalled();
  });

  it('uploads and sets the image when a valid file is selected', async () => {
    const mockImageUrl = 'https://example.com/image.jpg';
    mockHandleUploadImage.mockResolvedValueOnce(mockImageUrl);

    const { getByTestId } = render(
      <UploadImage onFileChange={mockOnFileChange} />,
    );

    const file = new File(['image content'], 'image.jpg', {
      type: 'image/jpeg',
    });
    const fileInput = getByTestId('file-input');

    await userEvent.upload(fileInput, file);

    await waitFor(() =>
      expect(mockHandleUploadImage).toHaveBeenCalledWith(file),
    );
    expect(mockOnFileChange).toHaveBeenCalledWith(mockImageUrl);
  });
});
