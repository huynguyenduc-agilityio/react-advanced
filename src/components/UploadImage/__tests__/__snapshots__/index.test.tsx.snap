// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadImage Component should match snapshot 1`] = `
<div>
  <div
    class="css-1dk8try"
  >
    <div
      class="chakra-form-control css-0"
      role="group"
    >
      <div
        class="css-p4mkuj"
      >
        <button
          aria-label="Upload image"
          class="chakra-button css-hv5zu9"
          type="button"
        >
          <svg
            fill="none"
            height="33"
            viewBox="0 0 33 33"
            width="33"
          >
            <circle
              cx="16.75"
              cy="16.5"
              fill="#CB3CFF"
              opacity="0.2"
              r="16"
            />
            <path
              d="M23.4167 8.5H10.0833C9.72971 8.5 9.39057 8.64048 9.14052 8.89052C8.89048 9.14057 8.75 9.47971 8.75 9.83333V20.5V23.1667C8.75 23.5203 8.89048 23.8594 9.14052 24.1095C9.39057 24.3595 9.72971 24.5 10.0833 24.5H23.4167C23.7703 24.5 24.1094 24.3595 24.3595 24.1095C24.6095 23.8594 24.75 23.5203 24.75 23.1667V9.83333C24.75 9.47971 24.6095 9.14057 24.3595 8.89052C24.1094 8.64048 23.7703 8.5 23.4167 8.5ZM21.025 15.1667C20.774 14.9193 20.4357 14.7807 20.0833 14.7807C19.7309 14.7807 19.3927 14.9193 19.1417 15.1667L15.4167 18.8917L13.6917 17.1667C13.4406 16.9193 13.1024 16.7807 12.75 16.7807C12.3976 16.7807 12.0594 16.9193 11.8083 17.1667L10.0833 18.8917V9.83333H23.4167V17.5583L21.025 15.1667Z"
              fill="#CB3CFF"
            />
          </svg>
        </button>
        <p
          class="chakra-text css-1f3jbo9"
        >
          <span
            class="chakra-text css-ynzqlz"
          >
            Click to upload
          </span>
           
          or drag and drop
        </p>
        <p
          class="chakra-text css-1q8knsm"
        >
          SVG, PNG, JPG or GIF (max. 800 x 400px)
        </p>
        <input
          accept="image/*"
          class="chakra-input css-0"
          data-testid="file-input"
          hidden=""
          id="imageInput"
          type="file"
        />
      </div>
    </div>
  </div>
</div>
`;
